R/pre_test_power.R

```diff
@@ -1997,3 +1506,614 @@ GeoLiftPower <- function(data,
+    )
+  }
+
+  # Part 1: Treatment and pre-treatment periods
+  data <- data %>% dplyr::rename(Y = paste(Y_id), location = paste(location_id), time = paste(time_id))
+  max_time <- max(data$time)
+  data$location <- tolower(data$location)
+  include_markets <- tolower(include_markets)
+  exclude_markets <- tolower(exclude_markets)
+
+  # Data Checks
+
+  # Small Pre-treatment Periods
+  if (max_time / max(treatment_periods) < 4) {
+    message(paste0("Caution: Small pre-treatment period!.
+                   \nIt's recommended to have at least 4x pre-treatment periods for each treatment period.\n"))
+  }
+
+  # Populate N if it's not provided
+  if (length(N) == 0) {
+    N <- unique(round(quantile(c(1:length(unique(data$location))),
+      probs = seq(0, 0.5, 0.1),
+      type = 1,
+      names = FALSE
+    )))
+
+    if (length(include_markets) > 0) {
+      # Keep only those equal or larger than included markets
+      N <- append(length(include_markets), N[length(include_markets) <= N])
+    }
+  }
+
+  # More include_markets than N
+  if (length(include_markets) > 0 & min(N) < length(include_markets)) {
+    message(paste0(
+      "Error: More forced markets than total test ones.",
+      " Consider increasing the values of N."
+    ))
+    return(NULL)
+  }
+
+  # Check that the provided markets exist in the data.
+  if (!all(tolower(include_markets) %in% tolower(unique(data$location)))) {
+    message(paste0(
+      "Error: One or more markets in include_markets were not",
+      " found in the data. Check the provided list and try again."
+    ))
+    return(NULL)
+  }
+
+  # Check that the provided markets exist in the data.
+  if (!all(tolower(exclude_markets) %in% tolower(unique(data$location)))) {
+    message(paste0(
+      "Error: One or more markets in exclude_markets were not",
+      " found in the data. Check the provided list and try again."
+    ))
+    return(NULL)
+  }
+
+  # Make sure all simulated effect sizes have the same sign.
+  if (min(effect_size < 0 & max(effect_size) > 0)) {
+    message(paste0(
+      "Error: The specified simulated effect sizes are not all of the same ",
+      " sign. \nTry again with a vector of all positive or negative effects",
+      " sizes that includes zero."
+    ))
+    return(NULL)
+  }
+
+  #  Check the holdout parameter
+  if (length(holdout) > 1) {
+    if (length(holdout) > 2) {
+      message("Error: Too many arguments in holdout. Provide the min and max holdout sizes.")
+      return(NULL)
+    } else if (holdout[1] >= holdout[2]) {
+      message("Error: The first argument in holdout should be strictly smaller than the second.")
+      return(NULL)
+    } else if (min(holdout) < 0 | max(holdout > 1)) {
+      message("Error: Please specify valid values for holdouts (values from 0 to 1)")
+      return(NULL)
+    }
+  } else if (length(holdout) == 1) {
+    message("Error: Too few arguments in holdout. Provide the min and max holdout sizes.")
+    return(NULL)
+  }
```

Comment:
Artur, what do you think about reducing the lines of code in our final function? I am thinking it could have 5 parts, I added the cases in which maybe its worth creating new mehods:

1. (NEW FUNCTION) Pre-processing and data checks.  Fail if checks are not passed, else continue.
2. (NEW FUNCTION) Power calculations using pvalue and loops (which is pretty similar to the one in GeoLiftPowerFinder, so we can probably unify right?).  Output should be a dataframe with the results of simulations.
3. (KEEP IN FUNCTION) Compute extra variables and clean simulated dataset based on input params.  Maybe we can keep this in the GeoLiftMarketSelection as its pretty specific.
4. (NEW FUNCTION) Rank function (which is also reused by GeoLiftPowerFinder).
5. (KEEP IN FUNCTION) Output for plots + simulations.

Reply:
Creating these new functions makes it easier if we ever want to create a new version of GeoLiftMarketSelection or if we want to reuse bits and pieces of it for other things.  What do you think?